### PR TITLE
feat(forms/datalist): be able to display labels

### PR DESCRIPTION
**Pull request by ncomont**
**What is the problem this PR is trying to solve?**
Impossible to display labels.

**What is the chosen solution to this problem?**
EnumOptions are used as label when availables.

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 1]
```diff
@@ -94,36 +32,34 @@ function defaultRenderNoMatch() {
 }
 
 /**
+ * Convert a datalist item as a value/label pair if needed
+ */
+function getValueLabelPair(item) {
+	return typeof item === 'object'
+	? { [item.value]: item.label }
+	: { [item]: item };
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Please use a regular if/else statement.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 2]
```diff
@@ -94,36 +32,34 @@ function defaultRenderNoMatch() {
...
+}
+
+/**
+ * Returns the value/label map from given items
+ */
+function getItemsMap(items) {
+	const itemsMap = {};
+	(items || []).forEach(item => Object.assign(itemsMap, getValueLabelPair(item)));
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Please return early in case of empty `items`. It avoids any processing.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 3]
```diff
@@ -203,41 +142,80 @@ class DatalistWidget extends React.Component {
 		}
 	}
 
+	getLabel(value) {
+		const { itemsMap } = this.state;
+
+		if (itemsMap && Object.keys(itemsMap).length) {
+			return Object.prototype.hasOwnProperty.call(itemsMap, value)
+				? itemsMap[value]
+				: value;
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Please avoid multi-line ternaries.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 4]
```diff
@@ -203,41 +142,80 @@ class DatalistWidget extends React.Component {
...
 
 	initSuggestions(value) {
-		let items;
-		if (this.props.schema.enum) {
-			items = this.props.schema.enum;
-		} else if (this.props.formContext.fetchItems) {
-			items = this.props.formContext.fetchItems(this.props.schema.title);
-		}
-		const suggestions = getMatchingSuggestions(items, value);
+		let keys = [];
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Why did you use a `let`?

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 5]
```diff
@@ -203,41 +142,80 @@ class DatalistWidget extends React.Component {
...
+		const regex = new RegExp(escapedValue, 'i');
+		return suggestions.filter(item => regex.test(this.getLabel(item)));
+	}
+
 	setValue(value) {
 		this.setState({ value });
 	}
 
+	getItems() {
+		const props = this.props;
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: If your goal is to write less code please spread the `props`.

Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: :ok_hand:

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 4 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: :ok_hand:

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 1 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: :cry:

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 6]
```diff
@@ -13,117 +13,69 @@ export function escapeRegexCharacters(str) {
...
 	}
 
-	return (
-		<div className={classnames(theme.item, 'datalist-item')}>
-			{emphasisedText.map((val, index) => <span key={index}>{val}</span>)}
-		</div>
-	);
+	const itemsMap = {};
+	items.forEach(item => Object.assign(itemsMap, getValueLabelPair(item)));
+	return itemsMap;
```
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: return items.reduce((a,b) => Object.assign({}, a, getValueLabelPair(b)), {}) ?

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 7]
```diff
@@ -135,7 +87,7 @@ class DatalistWidget extends React.Component {
 		this.inputProps = {
 			placeholder: props.placeholder,
 			required: props.required,
-			onBlur: event => this.onBlur(event),
+			onBlur: () => this.onBlur(),
```
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: question: why did you remove the event ?

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 8]
```diff
@@ -162,20 +117,21 @@ class DatalistWidget extends React.Component {
 		};
 	}
 
-	onBlur(event) {
+	componentWillMount() {
```
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: please use componentDidMount, this one is deprecated

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 9]
```diff
@@ -13,117 +13,69 @@ export function escapeRegexCharacters(str) {
...
-	return (
-		<div className={`${theme['items-container']} ${theme['no-result']}`}>
-			<span>No match.</span>
-		</div>
-	);
+function processValue(value) {
+	if (value && value.length) {
+		return value;
+	}
+	return undefined;
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Please don't return undefined.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 6 above)
Comment (acateland) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: object assign not really required since the reduced value is initialy a new object literal

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 10]
```diff
@@ -13,117 +13,69 @@ export function escapeRegexCharacters(str) {
...
 
-	return (
-		<div className={classnames(theme.item, 'datalist-item')}>
-			{emphasisedText.map((val, index) => <span key={index}>{val}</span>)}
-		</div>
-	);
+	const itemsMap = {};
+	items.forEach(item => Object.assign(itemsMap, getValueLabelPair(item)));
+	return itemsMap;
 }
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Why not start by defining the `itemsMap`. Return it empty in the if case or return it filed on the other case.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 11]
```diff
@@ -261,39 +254,122 @@ class DatalistWidget extends React.Component {
...
-		const renderItemsContainer =
-			this.props.renderItemsContainer || defaultRenderDatalistItemContainer;
+		let renderItemsContainer;
+		const value = this.state.value;
+		const renderItemData = { value };
 		const renderNoMatch = this.props.renderNoMatch || defaultRenderNoMatch;
+
+		this.inputProps.value = value;
+		if (this.state.noMatch) {
+			renderItemsContainer = renderNoMatch;
```
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: this assign const can be removed 
`renderItemsContainer = this.props.renderNoMatch || defaultRenderNoMatch`

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 7 above)
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Why did you remove the event?

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js [hunk 12]
```diff
@@ -162,20 +117,21 @@ class DatalistWidget extends React.Component {
...
-			this.props.onChange(this.state.value);
-			this.resetSuggestions();
-		} else {
-			const { value } = event.target;
-			if (value !== this.state.value) {
-				this.props.onChange(value);
-			}
+		} else if ((options.restricted && included) || !options.restricted) {
+			const { value } = this.state;
+			this.props.onChange(processValue(value));
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Why not use `this.state.value` directly? It remove one allocation.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 6 above)
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: good point so 
`return items.reduce((a,b) => Object.assign(a, getValueLabelPair(b)), {})`

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 7 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Because it is not used anymore. It was used like that : `const { value } = event.target;` ; but value is never defined.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 11 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: It can't without doing what it was done before this little refactoring. Before, we had :

```javascript
const renderItemsContainer = this.props.renderItemsContainer || defaultRenderDatalistItemContainer;
const renderNoMatch = this.props.renderNoMatch || defaultRenderNoMatch;

...

renderItemsContainer={this.state.noMatch ? renderNoMatch : renderItemsContainer}
```

I think it is clearer with this single if/else. What do you think about that ?

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 10 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: This part is replaced by the reduce suggested by @jmfrancois

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 9 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: @frassinier has done that for a good reason.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 7 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Because it is not used anymore. It was used like that : const { value } = event.target; ; but value is never defined.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 12 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: Oops, thanks.

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 6 above)
Comment (ncomont) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: :ok_hand:

File: packages/forms/src/widgets/DatalistWidget/DatalistWidget.js (hunk 9 above)
Comment (jmfrancois) [inline packages/forms/src/widgets/DatalistWidget/DatalistWidget.js]: this is our lint @talend-glorieux
